src/YDialogSpy.cc

```diff
@@ -278,70 +474,427 @@ void fillTree( YWidgetTreeItem * 		parent,
+}
+
+/**
+ * Delete the currently selected widget
+ */
+void YDialogSpyPrivate::deleteWidget()
+{
+    auto w = selectedWidget();
+    if (!w) return;
+
+    auto parent = w->parent();
+    if (!parent) return;
+
+    yuiMilestone() << "removing widget: " << w << std::endl;
+    parent->removeChild(w);
+
+    if ( w->isValid() )
+    {
+        delete w;
     }
+
+    // any other child left after the removal?
+    if (!parent->hasChildren())
+    {
+        // add an Empty widget to have a valid widget tree
+        // e.g. empty VBoxes are not allowed
+        YUI::widgetFactory()->createEmpty(parent);
+    }
+
+    targetDialogUpdated();
 }
 
+/**
+ * Helper method - Is the widget a VBox or Hbox?
+ * @param  widget the widget
+ * @return        true if the widget is a VBox or HBox
+ */
+bool isBox(const YWidget *widget)
+{
+    return dynamic_cast<const YLayoutBox *>(widget);
+}
 
-void YDialogSpy::showDialogSpy( YDialog * dialog )
+/**
+ * Helper method - Is the widget a VBox?
+ * @param  widget the widget
+ * @return        true if the widget is a VBox
+ */
+bool isVBox(const YWidget *widget)
+{
+    auto box = dynamic_cast<const YLayoutBox *>(widget);
+    return box && box->primary() == YD_VERT;
+}
+
+/**
+ * Move the selected widget up/left or down/right. The visual direction
+ * actually depends on the widget, it just moves the widget to the begining
+ * or the end of the container.
+ * @param true = up move to the begining (up/left), false = to the end (down/right)
+ */
+void YDialogSpyPrivate::moveSelected(Direction direction)
+{
+    auto target_widget = selectedWidget();
+    if (!target_widget) return;
+
+    auto parent = target_widget->parent();
+    if (!parent || !isBox(parent)) return;
+
+    if (direction == MOVE_UP)
+    {
+        // the first child cannot be moved further
+        if (target_widget == parent->firstChild()) return;
+
+        auto i = find( parent->childrenBegin(), parent->childrenEnd(), target_widget );
+        if (i != parent->childrenEnd())
+        {
+            // swap with the preceeding widget
+            // Note: use a temporary variable to not rely on the argument evaluation order!
+            auto second = i--;
+            std::swap(*second, *i);
+        }
+    }
+    else
+    // moving down
+    {
+        // the last child cannot be moved further to the end
+        if (target_widget == parent->lastChild()) return;
+
+        auto i = find( parent->childrenBegin(), parent->childrenEnd(), target_widget );
+        if (i != parent->childrenEnd())
+        {
+            // swap with the succeeding widget
+            // Note: use a temporary variable to not rely on the argument evaluation order!
+            auto second = i++;
+            std::swap(*second, *i);
+        }
+    }
+
+    targetDialogUpdated();
+}
+
+/**
+ * Generic handler for adding widgets
+ * @param type Type of the widget to add
+ */
+void YDialogSpyPrivate::addWidget(const std::string &type)
```

Comment:
Um, we could, but that might require some additional work (at least thinking about the API globally) as this would change the public API. I wanted to make my changes local and not touch the libyui public API if nor really needed.

So I'd keep it here for now.